docs/bots-guide.md

```diff
@@ -0,0 +1,192 @@
+
+Every bot is built upon this structure:
+```
+class FollowupHandler(object):
+    '''
+    A docstring documenting this bot.
+    '''
+
+    def usage(self):
+        return '''Your description of the bot'''
+
+    def triage_message(self, message):
+        #add your code here
+
+    def handle_message(self, message, client, state_handler):
+# add your code here
+
+handler_class = FollowupHandler
+```
+* The class name (in this case *FollowupHandler*) can be defined by you and should match the name of your bot. To register your bot's class, adjust the last line `handler_class = FollwupHandler` to match your class name.
+* Every bot needs to implement the functions
+ * `usage(self)`
+ * `triage_message(self, message)`
+ * `handle_message(self, message, client)`
+* These functions are documented in the [next section](#bot-API).
+
+## Bot API
+This sections documents the functions every bot needs to implement and the structure of the bot's config file.
+
+### usage(self)
+Is called by the system to obtain information about this bot.
+#### Arguments
+* self - the instance the method is called on.
+#### Return values
+* A string describing the bot's functionality
+#### Example implementation
+```
+def usage(self):
+    return '''
+        This plugin will allow users to flag messages
+        as being follow-up items.  Users should preface
+        messages with "@followup".
+        Before running this, make sure to create a stream
+        called "followup" that your API user can send to.
+        '''
+```
+
+### triage_message(self, message)
+Is called by the system when a message was sent.
+#### Arguments
+* self - the instance the method is called on.
+* message - an object containing information about the message, e.g.
+ * content - the content of the message
+ * content_type - the type of the content, e.g. *'text/x-markdown'* for normal messages
+ * display_recipient - the name of the stream the message is sent to (string)
+ * is_mentioned - is the bot pinged with an '@' in the message? (boolean)
+ * sender_email - email of the sender (string)
+ * sender_full_name - full name of the sender (string)
+ * subject - topic of the message (string)
+ * timestamp - when was the message sent (integer)
+
+ ####Return values
+ * True if the bot should react to this message
+ * False otherwise
+
+ ####Example implementation
+```
+def triage_message(self, message):
+original_content = message['content']
+    if message['display_recipient'] == 'followup':
+        return False
+    is_follow_up = (original_content.startswith('@followup') or
+                        original_content.startswith('@follow-up'))
+    return is_follow_up
+```
+
+### handle_message(self, message, client)
+Is called when `triage_message` returns true, handles user message.
+#### Arguments
+* self - the instance the method is called on.
+* message - an object containing information about the message, e.g.
```

Comment:
We may want to de-duplicate this by saying something like "* message - a dictionary describing a Zulip message (click [here] to see more about the message fields)